src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceMarginViewMargin.cs

```diff
@@ -0,0 +1,238 @@
+            _textView.ZoomLevelChanged += OnZoomLevelChanged;
+            _optionService.OptionChanged += OnRoslynOptionChanged;
+
+            _mainCanvas.Width = HeightAndWidthOfMargin;
+
+            _grid.LayoutTransform = new ScaleTransform(
+                scaleX: _textView.ZoomLevel / 100,
+                scaleY: _textView.ZoomLevel / 100);
+            _grid.LayoutTransform.Freeze();
+            UpdateMarginVisibility();
+        }
+
+        private void OnZoomLevelChanged(object sender, ZoomLevelChangedEventArgs e)
+        {
+            _grid.LayoutTransform = e.ZoomTransform;
+            _refreshAllGlyphs = true;
+        }
+
+        private void OnLayoutChanged(object sender, TextViewLayoutChangedEventArgs e)
+        {
+            _glyphManager.SetSnapshotAndUpdate(
+                _textView.TextSnapshot,
+                e.NewOrReformattedLines,
+                e.VerticalTranslation ? _textView.TextViewLines : e.TranslatedLines);
+
+            IList<ITextViewLine> lines = _refreshAllGlyphs ? _textView.TextViewLines : e.NewOrReformattedLines;
+            foreach (var line in lines)
+            {
+                _glyphManager.RemoveGlyph(line.Extent);
+                RefreshGlyphsOver(line);
+            }
+
+            _refreshAllGlyphs = false;
+        }
+
+        private void OnRoslynOptionChanged(object sender, OptionChangedEventArgs e)
+        {
+            if (e.Option == FeatureOnOffOptions.ShowInheritanceMargin || e.Option == FeatureOnOffOptions.InheritanceMarginCombinedWithIndicatorMargin)
+            {
+                UpdateMarginVisibility();
+            }
+        }
+
+        private void UpdateMarginVisibility()
+        {
+            var featureEnabled = _optionService.GetOption(FeatureOnOffOptions.ShowInheritanceMargin, _languageName) != false;
+            var showMargin = !_optionService.GetOption(FeatureOnOffOptions.InheritanceMarginCombinedWithIndicatorMargin, _languageName);
+            if (showMargin && featureEnabled)
+            {
+                _mainCanvas.Visibility = Visibility.Visible;
+            }
+            else
+            {
+                _mainCanvas.Visibility = Visibility.Collapsed;
+            }
+        }
+
+        private void OnTagsChanged(object sender, BatchedTagsChangedEventArgs e)
+        {
+            if (_textView.IsClosed)
+            {
+                return;
+            }
+
+            var changedSnapshotSpans = e.Spans.SelectMany(span => span.GetSpans(_textView.TextSnapshot)).ToImmutableArray();
+            if (changedSnapshotSpans.Length == 0)
+            {
+                return;
+            }
+
+            var startOfChangedSpan = changedSnapshotSpans.Min(span => span.Start);
+            var endOfChangedSpan = changedSnapshotSpans.Max(span => span.End);
+            var changedSpan = new SnapshotSpan(startOfChangedSpan, endOfChangedSpan);
+
+            _glyphManager.RemoveGlyph(changedSpan);
```

Comment:
is RemoveGlyph our own method?  if so, consider renaming to RemoveGlyphs as i presume that's what it is doing.